lib/min_heap.rb

```diff
@@ -1,77 +1,114 @@
 class HeapNode
   attr_reader :key, :value
-
+  
   def initialize(key, value)
     @key = key
     @value = value
   end
 end
 
 class MinHeap
-
+  
   def initialize
     @store = []
   end
-
+  
   # This method adds a HeapNode instance to the heap
-  # Time Complexity: ?
-  # Space Complexity: ?
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1)
   def add(key, value = key)
-    raise NotImplementedError, "Method not implemented yet..."
+    @store << HeapNode.new(key, value)
+    
+    if @store.length > 1
+      heap_up(@store.length - 1)
+    end
   end
-
+  
   # This method removes and returns an element from the heap
   #   maintaining the heap structure
-  # Time Complexity: ?
-  # Space Complexity: ?
+  # Time Complexity: O(log n)
+  # Space Complexity: O(1) It only stores the result variable, which is not altered by 
+  # the size of the input.
   def remove()
-    raise NotImplementedError, "Method not implemented yet..."
+    if @store.empty?
+      return nil
+    end
+    
+    swap(0, @store.length - 1)
+    result = @store.pop
+    
+    heap_down(0) unless @store.empty?
+    return result.value
   end
-
-
+  
+  
   # Used for Testing
   def to_s
     return "[]" if @store.empty?
-
+    
     output = "["
     (@store.length - 1).times do |index|
       output += @store[index].value + ", "
     end
-
+    
     output += @store.last.value + "]"
-      
+    
     return output
   end
-
+  
   # This method returns true if the heap is empty
-  # Time complexity: ?
-  # Space complexity: ?
+  # Time complexity: O(1) - arrays keep track of their length so that this can be contant time retrieval.
+  # Space complexity: O(1)
   def empty?
-    raise NotImplementedError, "Method not implemented yet..."
+    @store.empty?
   end
-
+  
   private
-
+  
   # This helper method takes an index and
   #  moves it up the heap, if it is less than it's parent node.
   #  It could be **very** helpful for the add method.
-  # Time complexity: ?
-  # Space complexity: ?
-  def heap_up(index)
+  # Time complexity: O(log n); it compares with parent nodes, not all nodes.
+  # Space complexity: O(1); swaps are happening in-place.
+  def heap_up(index) 
```

Comment:
Since this is a recursive method the space complexity is O(log n) because of the system stack!